game-backend/src/helper/gameObjects.ts

```diff
@@ -51,8 +51,10 @@ export const createUnit = ({
   type: GameObjectTypes.Unit,
 });
 
-const getObjectById = (state: State, id: ObjectId): GameObject =>
-  state.objectMap[id];
+export const getObjectById = <T extends GameObject>(
+  state: State,
+  id: ObjectId
+): T => state.objectMap[id] as T;
```

Comment:
Try not to type cast - this erases the benefit of TypeScript. Maybe adding an assertion here would make sense.